### PR TITLE
Conversation sampler

### PR DESCRIPTION
**Pull request by lucasgautheron**
- [x] implementation
- [x] docs
- [x] tests

### CODE REVIEW

**Comment by lucasgautheron** [comment]
@alecristia I made the High-vol sampler more "generic" by replace 'cvc', 'awc' and 'ctc' with 'vocs', 'words' and 'turns'.
Basically you can do as much as before but you get to choose the speakers to target for each metric.
Is this fine with you?

**Comment by alecristia** [comment]
> @alecristia I made the High-vol sampler more "generic" by replace 'cvc', 'awc' and 'ctc' with 'vocs', 'words' and 'turns'. Basically you can do as much as before but you get to choose the speakers to target for each metric. Is this fine with you?

YES! let me take a quick look...

File: ChildProject/pipelines/samplers.py [hunk 1]
```diff
@@ -645,7 +645,7 @@ class HighVolubilitySampler(Sampler):
     :type project: ChildProject.projects.ChildProject
     :param annotation_set: set of annotations to calculate volubility from.
     :type annotation_set: str
-    :param metric: the metric to evaluate high-volubility. should be any of 'awc', 'ctc', 'cvc'.
+    :param metric: the metric to evaluate high-volubility. should be any of 'words', 'turns', 'vocs'.
```
Comment (alecristia) [inline ChildProject/pipelines/samplers.py:648]: hmm reading it like this I think LENA users will be confused. Perhaps words (awc), turns (ctc), vocs (cvc)?

File: ChildProject/pipelines/samplers.py [hunk 2]
```diff
@@ -711,74 +713,65 @@ def _segment_scores(self, recording):
...
             else:
                 segments.sort_values(["segment_onset", "segment_offset"])
-                segments = segments[
-                    segments["speaker_type"].isin(["FEM", "MAL", "CHI"])
-                ]
-                segments["duration"] = (
-                    segments["segment_offset"] - segments["segment_onset"]
-                )
+                segments = segments[segments["speaker_type"].isin(self.speakers)]
                 segments["iti"] = segments["segment_onset"] - segments[
```
Comment (alecristia) [inline ChildProject/pipelines/samplers.py:725]: Right, so as to leave a note on the logic:
- for LENA, we just follow LENA's counts
- for ACLEW, we select turns concerning the speakers of choice

File: ChildProject/pipelines/samplers.py [hunk 3]
```diff
@@ -711,74 +713,65 @@ def _segment_scores(self, recording):
...
+                segments["is_CT"] = (segments["iti"] < 1000) & (
+                    (
+                        (segments["speaker_type"] == "CHI")
+                        & (segments["prev_speaker_type"].isin(key_child_environment))
+                    )
+                    | (
+                        (segments["speaker_type"].isin(key_child_environment))
+                        & (segments["prev_speaker_type"] == "CHI")
+                    )
                 )
```
Comment (alecristia) [inline ChildProject/pipelines/samplers.py:741]: simplification of lambda/axis - no negative/different behavior consequences that you can think of, right?

Comment (alecristia) [reply ChildProject/pipelines/samplers.py:741]: (oops, that was intended as a comment, not a review item)

File: ChildProject/pipelines/samplers.py [hunk 4]
```diff
@@ -711,74 +713,65 @@ def _segment_scores(self, recording):
...
             else:
                 segments = (
-                    segments[segments.speaker_type == "CHI"]
+                    segments[segments.speaker_type.isin(self.speakers)]
                     .groupby("chunk", as_index=False)[["segment_onset"]]
                     .count()
-                    .rename(columns={"segment_onset": "cvc"})
+                    .rename(columns={"segment_onset": self.metric})
                     .merge(windows)
                 )
```
Comment (alecristia) [inline ChildProject/pipelines/samplers.py:768]: very elegant!

File: docs/source/samplers.rst [hunk 5]
```diff
@@ -86,3 +86,22 @@ High-Volubility sampler
...
 
    child-project sampler /path/to/dataset /path/to/destination high-volubility --help
+
+Conversation sampler
+~~~~~~~~~~~~~~~~~~~~
+
+The conversation sampler returns the conversational blocks with the highest amount of turns.
+The first step is the detection of conversational blocks.
+Two consecutive vocalizations are considered part of the same conversational block if they are not separated
+by an interval longer than a certain duration, which by default is set to 1000 milliseconds.
```
Comment (alecristia) [inline docs/source/samplers.rst:96]: should we perhaps clarify that this is only for the ACLEW pipeline? if I read the code correctly above, the filter for 1k ms is not applied to subset from the TIFR/TIMR

File: docs/source/samplers.rst [hunk 6]
```diff
@@ -86,3 +86,22 @@ High-Volubility sampler
...
+
+Then, the amount of conversational turns (between the key child and adult) is calculated for each conversational block.
+The sampler returns, for each unit, the desired amount of conversations with the higher amount of turns.
```
Comment (alecristia) [inline docs/source/samplers.rst:99]: ah, I read the code above as suggesting that I can select only FEM-CHI blocks -- did I get that wrong?

File: ChildProject/pipelines/samplers.py (hunk 3 above)
Comment (lucasgautheron) [reply ChildProject/pipelines/samplers.py:746]: nope, if anything it should be faster!

File: docs/source/samplers.rst (hunk 5 above)
Comment (lucasgautheron) [reply docs/source/samplers.rst:102]: done!

File: docs/source/samplers.rst (hunk 6 above)
Comment (lucasgautheron) [reply docs/source/samplers.rst:105]: you can limit the search to FEM vs CHI only, but by default the search includes all both FEM and MAL. It is more clear now ?

File: ChildProject/pipelines/samplers.py (hunk 2 above)
Comment (lucasgautheron) [reply ChildProject/pipelines/samplers.py:730]: Now mentioned in the docs

File: ChildProject/pipelines/samplers.py (hunk 1 above)
Comment (lucasgautheron) [reply ChildProject/pipelines/samplers.py:654]: I tried to make it more clear in the docs (both the API and the cli)

File: docs/source/samplers.rst (hunk 6 above)
Comment (alecristia) [reply docs/source/samplers.rst:105]: Then, the amount of conversational turns (between the key child and adult) --> Then, the amount of conversational turns (by default, between the key child and female/male adult)

**Pull request merged**